File: app/domain/models/bill_analysis.py
"""
Bill Analysis presentation models — Integrum Energy production module.

These DTOs shape API responses for the Bill Analysis UI.
They do not replace domain models; they describe what the client should render.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


BillAnalysisStatus = Literal[
    "needs_review",
    "ready",
    "unsupported",
    "error",
]

ConfidenceLevel = Literal["HIGH", "MEDIUM", "LOW", "MISSING"]


class FieldAuditEntry(BaseModel):
    field: str
    original_value: Any | None = None
    corrected_value: Any | None = None
    corrected_by_user: bool = True
    corrected_at: datetime
    action: Literal["corrected", "accepted_as_printed", "category_confirmed"]


class BillFieldView(BaseModel):
    name: str
    label: str
    section: str
    value: Any | None = None
    display_value: str
    confidence: float = 0.0
    level: ConfidenceLevel = "MISSING"
    source: str = "unknown"
    needs_verification: bool = False
    editable: bool = True
    required: bool = False


class BillSectionView(BaseModel):
    id: str
    title: str
    fields: list[BillFieldView] = Field(default_factory=list)


class SupportView(BaseModel):
    supported: bool
    state: str = "Karnataka"
    discom: str | None = None
    category: str | None = None
    is_bescom_bill: bool | None = None
    can_analyze: bool = False
    message: str
    block_reasons: list[str] = Field(default_factory=list)


class ValidationIssueView(BaseModel):
    code: str
    message: str
    field: str | None = None
    severity: str = "WARNING"


class BillCalculationView(BaseModel):
    units_consumed: float | None = None
    total_amount: float | None = None
    cost_per_unit: float | None = None
    charge_lines_sum: float | None = None
    charge_total_delta: float | None = None
    annualized_units_estimate: float | None = None
    annualized_amount_estimate: float | None = None
    notes: list[str] = Field(default_factory=list)


class HistoryBillView(BaseModel):
    analysis_id: str
    billing_period: str | None = None
    bill_date: str | None = None
    units_consumed: float | None = None
    total_amount: float | None = None


class HistorySummaryView(BaseModel):
    consumer_id: str | None = None
    bill_count: int = 0
    ready_for_trend_analysis: bool = False
    bills: list[HistoryBillView] = Field(default_factory=list)
    duplicate_warnings: list[str] = Field(default_factory=list)


class BillAnalysisView(BaseModel):
    analysis_id: str
    status: BillAnalysisStatus
    message: str
    document: dict[str, Any] = Field(default_factory=dict)
    sections: list[BillSectionView] = Field(default_factory=list)
    support: SupportView
    validation_issues: list[ValidationIssueView] = Field(default_factory=list)
    consistency_warnings: list[str] = Field(default_factory=list)
    needs_confirmation: list[str] = Field(default_factory=list)
    calculations: BillCalculationView | None = None
    history: HistorySummaryView | None = None
    corrections_audit: list[FieldAuditEntry] = Field(default_factory=list)
    confirmed: bool = False

File: frontend/lib/bill-analysis.ts
export type ConfidenceLevel = "HIGH" | "MEDIUM" | "LOW" | "MISSING";

export type BillAnalysisStatus =
  | "needs_review"
  | "ready"
  | "unsupported"
  | "error";

export type BillField = {
  name: string;
  label: string;
  section: string;
  value: string | number | boolean | null;
  display_value: string;
  confidence: number;
  level: ConfidenceLevel;
  source: string;
  needs_verification: boolean;
  editable: boolean;
  required?: boolean;
};

export type BillSection = {
  id: string;
  title: string;
  fields: BillField[];
};

export type SupportInfo = {
  supported: boolean;
  state: string;
  discom: string | null;
  category: string | null;
  is_bescom_bill: boolean | null;
  can_analyze: boolean;
  message: string;
  block_reasons: string[];
};

export type BillCalculation = {
  units_consumed: number | null;
  total_amount: number | null;
  cost_per_unit: number | null;
  charge_lines_sum: number | null;
  charge_total_delta: number | null;
  annualized_units_estimate: number | null;
  annualized_amount_estimate: number | null;
  notes: string[];
};

export type HistoryBill = {
  analysis_id: string;
  billing_period: string | null;
  bill_date: string | null;
  units_consumed: number | null;
  total_amount: number | null;
};

export type HistorySummary = {
  consumer_id: string | null;
  bill_count: number;
  ready_for_trend_analysis: boolean;
  bills: HistoryBill[];
  duplicate_warnings: string[];
};

export type BillAnalysis = {
  analysis_id: string;
  status: BillAnalysisStatus;
  message: string;
  document: Record<string, unknown>;
  sections: BillSection[];
  support: SupportInfo;
  validation_issues: { code: string; message: string; field?: string; severity: string }[];
  consistency_warnings: string[];
  needs_confirmation: string[];
  calculations: BillCalculation | null;
  history: HistorySummary | null;
  corrections_audit: unknown[];
  confirmed: boolean;
};

export type ExtractResponse = {
  message: string;
  analysis: BillAnalysis;
};

export type BatchItem = {
  filename: string;
  status: string;
  analysis_id?: string;
  billing_period?: string | null;
  units_consumed?: number | null;
  total_amount?: number | null;
  needs_confirmation?: string[];
  duplicate_warnings?: string[];
  error?: string;
  analysis?: BillAnalysis;
};

export type BatchResponse = {
  message: string;
  processed: number;
  successful: number;
  needs_review: number;
  failed: number;
  items: BatchItem[];
};

export type ProcessingStep =
  | "idle"
  | "uploading"
  | "reading"
  | "extracting"
  | "validating"
  | "checking"
  | "review"
  | "ready"
  | "error";

export const PROCESSING_LABELS: Record<ProcessingStep, string> = {
  idle: "",
  uploading: "Uploading…",
  reading: "Reading bill…",
  extracting: "Extracting details…",
  validating: "Validating…",
  checking: "Checking BESCOM eligibility…",
  review: "Ready for review",
  ready: "Analysis ready",
  error: "Processing failed",
};

export function formatCurrency(value: number | null | undefined): string {
  if (value == null) return "—";
  return `₹${value.toLocaleString("en-IN", { maximumFractionDigits: 2 })}`;
}

export function formatUnits(value: number | null | undefined): string {
  if (value == null) return "—";
  return `${value.toLocaleString("en-IN", { maximumFractionDigits: 1 })} kWh`;
}

export function maskAccount(value: string | null | undefined): string {
  if (!value) return "";
  if (value.length <= 4) return value;
  return `${"*".repeat(Math.max(0, value.length - 4))}${value.slice(-4)}`;
}
